File: js/16401.js
const fs = require('fs');

const inputs = fs
  .readFileSync('example.txt')
  .toString()
  .split('\n')
  .map((e) => e.split(' ').map(Number));

const [M, N] = inputs[0];
const arr = inputs[1];

arr.sort((a, b) => a - b);

let start = 1;
let end = arr[arr.length - 1];

while (start <= end) {
  const mid = Math.floor((start + end) / 2);

  if (isPossilbe(mid)) {
    start = mid + 1;
  } else {
    end = mid - 1;
  }
}

console.log(end);

function isPossilbe(k) {
  let sum = 0;
  for (let i = 0; i < N; i++) {
    sum += Math.floor(arr[i] / k);
  }

  return sum >= M;
}

File: c++/16401.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

vector<int> v;
int M, N;

bool isPossible(int k) {
  int sum = 0;
  for (int i = 0; i < N; i++) {
    sum += v[i] / k;
  }

  return sum >= M;
}

int main() {
  freopen("example.txt", "r", stdin);

  cin >> M >> N;

  for (int i = 0; i < N; i++) {
    int a;
    cin >> a;
    v.push_back(a);
  }

  sort(v.begin(), v.end());

  int start = 1;
  int end = v.back();

  while (start <= end) {
    int mid = (start + end) / 2;

    if (isPossible(mid)) {
      start = mid + 1;
    } else {
      end = mid - 1;
    }
  }

  cout << end;

  return 0;
}